Only validate expiration date of incoming token if token_ttl is set

pluk77 opened the pull request:
fix #889

pluk77 committed:
Only validate expiration date of incoming token if token_ttl is set

fix #889

pluk77 committed:
Updated result of isExpired with empty payload

isExpired should not return false if the token is empty. As it infers that the ttl of an empty token is set and in the future.

pluk77 committed:
isExpired on an empty payload should return false

As it is valid to have a payload without an exp, the test should return false as a token only expires if there is an exp date set.

pluk77 committed:
Empty payloads should get an invalid state

Although proper validation is probably better, ensuring an empty payload is invalid is the first step

pluk77 commented:
Is there a specific reason this patch has not been reviewed?

chalasr commented:
Only that my open-source time is limited. Thanks for the ping, this looks good to me, I need to check if it doesn't have "security" implications (e.g. that it does not open the door for tokens without `exp` while `token_ttl` is configured, etc.).
I'll be back on this as soon as I can.

pluk77 commented:
@chalasr Thank you. I fully understand and appreciate your time constraints. If you do find security implications, please let me know what they are so we can find a better solution for the issue at hand.

chalasr commented:
Apologies again for the delay, this is a long standing topic.
Looking at this twice, I think this deserves a config option. It could be e.g. `verify_tokens_without_exp` with `false` as default. That setting would then be injected in the jws provider and passed to the `LoadedJWS` VO somehow to accept tokens with no `exp` claim when the option is set to true.
Would you mind updating your PR in that direction?
IMHO In production this only makes sense with a proper token invalidation mechanism, which I think we should add at some point (some storage abstraction alongside a revocation logic based on the `jti` claim).

pluk77 commented:
I will have a look and amend my PR accordingly.

pluk77 committed:
Update Configuration.php

Added validate_tokens_without_ttl

pluk77 committed:
Update LexikJWTAuthenticationExtension.php

Added validate_tokens_without_ttl

pluk77 committed:
Update deprecated.xml

Added validate_tokens_without_ttl

pluk77 committed:
Update deprecated_51.xml

added validate_tokens_without_ttl

pluk77 committed:
Update lcobucci.xml

Added validate_tokens_without_ttl

pluk77 committed:
Update DefaultJWSProvider.php

Added validateTokensWithoutTtl

pluk77 committed:
Update LcobucciJWSProvider.php

Added validateTokensWithoutTtl

pluk77 committed:
Update LoadedJWS.php

Added allowEmptyTtl

pluk77 committed:
Update LexikJWTAuthenticationExtensionTest.php

Added validate_tokens_without_ttl

pluk77 committed:
Update AbstractJWSProviderTest.php

Corrected ttl test

pluk77 committed:
Update LoadedJWS.php

added unknown state to prevent being stateless

pluk77 committed:
Update LoadedJWSTest.php

Updated test to allow for no ttl

pluk77 committed:
Update LoadedJWSTest.php

missing ;

pluk77 committed:
Update AbstractJWSProviderTest.php

Fix test

chalasr commented:
@pluk77 Do you want some help to fix the broken tests?

pluk77 commented:
@chalasr Yes please!! That would be awesome.

pluk77 commented:
Anybody able and willing to assist with getting these broken tests fixed so this PR can be merged?

pluk77 committed:
Update LcobucciJWSProvider.php

Allow for the creation of tokens without exp by setting it to 0 in the payload during creation

pluk77 committed:
Update DefaultJWSProvider.php

Allow for the creation of tokens without exp by setting it to 0 in the payload during creation

chalasr committed:
Tweak implementation

chalasr commented:
Just pushed on your branch, PR is now green :) Can you please confirm that the new option still works as expected?

pluk77 commented:
Thank you so much for your assistance here. I will go through the changes, but at first glance I think there are some changes which I do not understand the reasoning behind as they seem to contradict each other and some that will most likely still not allow for the different use-cases. I will run some tests and provide proper feedback and a potential solution for you to consider.

chalasr commented:
To make sure we are aligned, the goal of this PR is to allow validating tokens that have no `exp` claim when the option is set to `true`, right? Looking forward to your feedback.

pluk77 committed:
Update Configuration.php

set default of validate_tokens_without_ttl to true

pluk77 committed:
Update LcobucciJWSProvider.php

Set default to true to ensure backwards compatibility and improve security. Decoupled ttl setting from validation requirement ad token should be validated based on its content and signature, not on the ttl in the setting which is used during generation.

pluk77 committed:
Update DefaultJWSProvider.php

Make provider signatures the same

pluk77 committed:
Update DefaultJWSProvider.php

Correct ttl validation logic

pluk77 committed:
Update AbstractJWSProviderTest.php

correct exp validation test

pluk77 commented:
I have made changes to your update to ensure the default settings are secure and that the ttl setting is not over-ruling the newly added validation setting. 

My reasoning is that the ttl in the settings is the default setting used to generate a token. But tokens can also be generated via commands or code and the EXP can be set via the payload (I have submitted a different PR to ensure that works correctly)

For that reason, the ttl setting should not influence if tokens without an exp should be validated or not as the validator does not know if those tokens were generated with custom or blank EXP. Ideally, the signature of a token should be used to determine / indicate if the EXP was removed from a token.

It works on my application and most test pass, but seeing that the default setting is now flagging tokens without EXP as invalid, there are a few test that fail.

I am not sure where in the test to add the new 'validate_tokens_without_ttl: false' setting as I am sure that will make these tests pass.

pluk77 commented:
> To make sure we are aligned, the goal of this PR is to allow validating tokens that have no `exp` claim when the option is set to `true`, right? Looking forward to your feedback.

I understood it to be the other way around as this lines up with '$shouldCheckExpiration' in the LoadedJWS

No EXP + validate_tokens_without_ttl=false => results in token being validated without looking at EXP ($shouldCheckExpiration = false)
No EXP + validate_tokens_without_ttl=true => results in invalid token ($shouldCheckExpiration = true)
EXP + validate_tokens_without_ttl=false => results in EXP being validated ($shouldCheckExpiration = true)
EXP + validate_tokens_without_ttl=true=> results in EXP being validated ($shouldCheckExpiration = true)

chalasr commented:
Thanks for the explanation, I'll check this again asap this week.
I didn't look into your last changes yet, but please avoid changing existing tests to make sure they keep passing as they are. (The current behavior must not change to preserve backward compatibility, only new tests should be added to test the new behavior)

chalasr commented on Services/JWSProvider/DefaultJWSProvider.php:
```diff
@@ -45,16 +45,22 @@ class DefaultJWSProvider implements JWSProviderInterface
+    /**
+     * @var bool
+     */
+    private $validateTokensWithoutTtl;
```
This class should not get new features, it is deprecated. Only LcobucciJWSProvider should benefit from this change

pluk77 replied:
> This class should not get new features, it is deprecated. Only LcobucciJWSProvider should benefit from this change

Understood. I have reverted all changes to DefaultJWSProvider. The additional parameter in the constructor of LcobucciJWSProvider should not cause an issue as the constructor is not defined in the providerInterface.

pluk77 committed:
Update DefaultJWSProvider.php

Revert changes to deprecated class

pluk77 commented:
> I understood it to be the other way around as this lines up with '$shouldCheckExpiration' in the LoadedJWS

Perhaps we should change the setting name to something less confusing seeing that we both had a different understanding of the meaning? 

How about : validate_tokens_without_ttl_as_invalid = true (Default behavior, regardless of ttl value in yml file)

pluk77 commented:
Please consider my new PR #1020
